### ISSUE TITLE
Update the docker setup

### ISSUE THREAD

**Issue by Mnuzz** [Bit-Nation/BITNATION-Pangea-libs#55]
From Florian:

> @Mark would you mind updateing the docker setup?
> @Mark https://github.com/realm/realm-js/issues/1635

The docker setup currently doesn't run properly. Docker behaves the same way, regardless of the host operating system. So I am able to use my main dev machine for this (which isn't a Mac).

**Comment by Mnuzz** [comment]
I updated the node version in the docker image to the same as what I previously got to work on my Mac. The old version was 6.11.5. The new version is 8.9.4. I used a clean clone of the repository, on the latest `develop` commit (`830c715bc3dcdae9d7e7672ca37aeda1a147fef2`), and ran no commands except the following:
```
docker-compose up -d  
docker-compose exec node bash

(within docker image:)
npm install  
npm test  
```

Unfortunately, I get a new set of errors on Docker which I do not get on the Mac under the same test conditions:
```
 PASS  src/ethereum/utils.test.js
 FAIL  src/ethereum/nation.test.js
  ● nation › nations

    Failed: [object Object]

      at stackFormatter (node_modules/jest-jasmine2/build/expectation_result_factory.js:49:427)
          at <anonymous>

  ● nation › saveDraft › success

    Operation not permitted

      at Promise (node_modules/realm/lib/extensions.js:67:29)
          at new Promise (<anonymous>)
      at Function.open (node_modules/realm/lib/extensions.js:66:27)
      at dbFactory (src/database/db.js:39:135)
      at Object.done (src/ethereum/nation.test.js:225:41)
          at new Promise (<anonymous>)
          at <anonymous>

  ● nation › saveDraft › fail

    expect(received).toBe(expected)

    Expected value to be (using ===):
      {}
    Received:
      undefined

    Difference:

      Comparing two different types of values. Expected object but received undefined.

      at db.query.then.then.catch.response (src/ethereum/nation.test.js:279:41)
          at <anonymous>

  ● nation › updateDraft › no nation found

    Operation not permitted

      at Promise (node_modules/realm/lib/extensions.js:67:29)
          at new Promise (<anonymous>)
      at Function.open (node_modules/realm/lib/extensions.js:66:27)
      at dbFactory (src/database/db.js:39:135)
      at Object.done (src/ethereum/nation.test.js:288:41)
          at new Promise (<anonymous>)
          at <anonymous>

  ● nation › updateDraft › updated successfully

    Failed: [object Object]

      at stackFormatter (node_modules/jest-jasmine2/build/expectation_result_factory.js:49:427)
          at <anonymous>
      at runMicrotasksCallback (internal/process/next_tick.js:121:5)
      at _combinedTickCallback (internal/process/next_tick.js:131:7)
      at process._tickCallback (internal/process/next_tick.js:180:9)

  ● nation › updateDraft › nation update (db error)

    expect(received).toBe(expected)

    Expected value to be (using ===):
      "system_error.db_write_failed"
    Received:
      {"nation": {"decisionMakingProcess": "dictatorship", "diplomaticRecognition": false, "exists": true, "governanceService": "Security", "lawEnforcementMechanism": "xyz", "nationCode": "Code civil", "nationDescription": "We <3 cryptography", "nationName": "Bitnation", "nonCitizenUse": false, "profit": true, "virtualNation": false}, "transKey": "nation.draft.saved_failed"}

    Difference:

      Comparing two different types of values. Expected string but received object.

      at nations.saveDraft.then.catch.e (src/ethereum/nation.test.js:395:27)
          at <anonymous>
      at runMicrotasksCallback (internal/process/next_tick.js:121:5)
      at _combinedTickCallback (internal/process/next_tick.js:131:7)
      at process._tickCallback (internal/process/next_tick.js:180:9)

  ● nation › submitDraft › success

    Failed: [object Object]

      at stackFormatter (node_modules/jest-jasmine2/build/expectation_result_factory.js:49:427)
          at <anonymous>
      at runMicrotasksCallback (internal/process/next_tick.js:121:5)
      at _combinedTickCallback (internal/process/next_tick.js:131:7)
      at process._tickCallback (internal/process/next_tick.js:180:9)

  ● nation › saveAndSubmit › success

    Failed: [object Object]

      at stackFormatter (node_modules/jest-jasmine2/build/expectation_result_factory.js:49:427)
          at <anonymous>
      at runMicrotasksCallback (internal/process/next_tick.js:121:5)
      at _combinedTickCallback (internal/process/next_tick.js:131:7)
      at process._tickCallback (internal/process/next_tick.js:180:9)

  ● nation › deleteDraft › success

    Error
      [object Object] thrown

 FAIL  src/ethereum/wallet.test.js
  ● wallet › ethSync › success

    Operation not permitted

      at Promise (node_modules/realm/lib/extensions.js:67:29)
          at new Promise (<anonymous>)
      at Function.open (node_modules/realm/lib/extensions.js:66:27)
      at dbFactory (src/database/db.js:39:135)
      at Object.done (src/ethereum/wallet.test.js:233:41)
          at new Promise (<anonymous>)
          at <anonymous>

(node:502) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 38): Error: Operation not permitted
(node:502) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
(node:502) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 40): Error: Operation not permitted
 FAIL  src/profile/profile.test.js
  ● profile › setProfile › create profile

    ReferenceError: rej is not defined

      at p.setProfile.then.then.catch.err (src/profile/profile.test.js:57:26)
          at <anonymous>
      at process._tickCallback (internal/process/next_tick.js:188:7)

  ● profile › setProfile › try to update profile

    Operation not permitted

      at Promise (node_modules/realm/lib/extensions.js:67:29)
          at new Promise (<anonymous>)
      at Function.open (node_modules/realm/lib/extensions.js:66:27)
      at dbFactory (src/database/db.js:39:135)
      at Object.done (src/profile/profile.test.js:64:41)
          at new Promise (<anonymous>)
          at <anonymous>

  ● profile › getProfile › get profile that exist

    Operation not permitted

      at Promise (node_modules/realm/lib/extensions.js:67:29)
          at new Promise (<anonymous>)
      at Function.open (node_modules/realm/lib/extensions.js:66:27)
      at dbFactory (src/database/db.js:39:135)
      at Object.done (src/profile/profile.test.js:117:41)
          at new Promise (<anonymous>)
          at <anonymous>

  ● profile › hasProfile › false

    expect(received).resolves.toBeFalsy()

    Expected received Promise to resolve, instead it rejected to value
      [Error: Operation not permitted]

      at Object.<anonymous> (node_modules/expect/build/index.js:131:11)
          at Generator.throw (<anonymous>)
      at step (node_modules/expect/build/index.js:40:727)
      at node_modules/expect/build/index.js:40:926
          at <anonymous>
      at runMicrotasksCallback (internal/process/next_tick.js:121:5)
      at _combinedTickCallback (internal/process/next_tick.js:131:7)
      at process._tickCallback (internal/process/next_tick.js:180:9)

 FAIL  src/queues/transaction.test.js
  ● transaction › addJob › write job to database

    Operation not permitted

      at Promise (node_modules/realm/lib/extensions.js:67:29)
          at new Promise (<anonymous>)
      at Function.open (node_modules/realm/lib/extensions.js:66:27)
      at dbFactory (src/database/db.js:39:135)
      at Object.done (src/queues/transaction.test.js:32:41)
          at new Promise (<anonymous>)
          at <anonymous>

  ● transaction › process › processor not found

    Operation not permitted

      at Promise (node_modules/realm/lib/extensions.js:67:29)
          at new Promise (<anonymous>)
      at Function.open (node_modules/realm/lib/extensions.js:66:27)
      at dbFactory (src/database/db.js:39:135)
      at Object.test (src/queues/transaction.test.js:77:37)
          at new Promise (<anonymous>)

  ● transaction › process › processor not found

    expect(received).toBe(expected)

    Expected value to be (using ===):
      "Couldn't find processor for my_processor"
    Received:
      "Operation not permitted"

      at txQueue.addJob.then.then.catch.e (src/queues/transaction.test.js:116:35)
          at <anonymous>
      at process._tickCallback (internal/process/next_tick.js:188:7)

  ● transaction › process › added success message when job is completed

    Operation not permitted

      at Promise (node_modules/realm/lib/extensions.js:67:29)
          at new Promise (<anonymous>)
      at Function.open (node_modules/realm/lib/extensions.js:66:27)
      at dbFactory (src/database/db.js:39:135)
      at Object.done (src/queues/transaction.test.js:122:41)
          at new Promise (<anonymous>)
          at <anonymous>

 PASS  src/ethereum/PangeaProvider.test.js
 FAIL  src/queues/messaging.test.js
  ● messaging › addJob › save to database

    Operation not permitted

      at Promise (node_modules/realm/lib/extensions.js:67:29)
          at new Promise (<anonymous>)
      at Function.open (node_modules/realm/lib/extensions.js:66:27)
      at dbFactory (src/database/db.js:39:135)
      at Object.done (src/queues/messaging.test.js:71:41)
          at new Promise (<anonymous>)
          at <anonymous>

  ● messaging › fetchMessages › fetch limited amount messages

    Operation not permitted

      at Promise (node_modules/realm/lib/extensions.js:67:29)
          at new Promise (<anonymous>)
      at Function.open (node_modules/realm/lib/extensions.js:66:27)
      at dbFactory (src/database/db.js:39:135)
      at Object.done (src/queues/messaging.test.js:95:41)
          at new Promise (<anonymous>)
          at <anonymous>

 PASS  src/ethereum/web3.test.js
 PASS  src/test_implementations/nodeJsSecureStorage.test.js
 FAIL  src/database/db.test.js
  ● write › successfully

    expect(received).resolves.toBe()

    Expected received Promise to resolve, instead it rejected to value
      [Error: Operation not permitted]

      at Object.<anonymous> (node_modules/expect/build/index.js:131:11)
          at Generator.throw (<anonymous>)
      at step (node_modules/expect/build/index.js:40:727)
      at node_modules/expect/build/index.js:40:926
          at <anonymous>

  ● query › successfully

    expect(received).resolves.toEqual()

    Expected received Promise to resolve, instead it rejected to value
      [Error: Operation not permitted]

      at Object.<anonymous> (node_modules/expect/build/index.js:131:11)
          at Generator.throw (<anonymous>)
      at step (node_modules/expect/build/index.js:40:727)
      at node_modules/expect/build/index.js:40:926
          at <anonymous>

 PASS  src/database/queries.test.js

Test Suites: 6 failed, 5 passed, 11 total
Tests:       21 failed, 73 passed, 94 total
Snapshots:   0 total
Time:        6.459s
Ran all test suites.
```

**Comment by Mnuzz** [comment]
updating `package.json` and setting `realm` to `^2.2.4`, then running `npm install` and `npm test` created more failures  than before:

```
 FAIL  src/ethereum/nation.test.js
  ● nation › nations

    Operation not permitted

      at Function.open (node_modules/realm/lib/extensions.js:67:47)
      at dbFactory (src/database/db.js:39:135)
      at Object.done (src/ethereum/nation.test.js:32:64)
          at new Promise (<anonymous>)
          at <anonymous>

  ● nation › saveDraft › success

    Operation not permitted

      at Function.open (node_modules/realm/lib/extensions.js:67:47)
      at dbFactory (src/database/db.js:39:135)
      at Object.done (src/ethereum/nation.test.js:225:41)
          at new Promise (<anonymous>)
          at <anonymous>

  ● nation › saveDraft › fail

    Operation not permitted

      at Function.open (node_modules/realm/lib/extensions.js:67:47)
      at dbFactory (src/database/db.js:39:135)
      at Object.done (src/ethereum/nation.test.js:261:41)
          at new Promise (<anonymous>)
          at <anonymous>

  ● nation › updateDraft › no nation found

    Operation not permitted

      at Function.open (node_modules/realm/lib/extensions.js:67:47)
      at dbFactory (src/database/db.js:39:135)
      at Object.done (src/ethereum/nation.test.js:288:41)
          at new Promise (<anonymous>)
          at <anonymous>

  ● nation › updateDraft › updated successfully

    Operation not permitted

      at Function.open (node_modules/realm/lib/extensions.js:67:47)
      at dbFactory (src/database/db.js:39:135)
      at Object.done (src/ethereum/nation.test.js:302:41)
          at new Promise (<anonymous>)
          at <anonymous>

  ● nation › updateDraft › nation update (db error)

    Operation not permitted

      at Function.open (node_modules/realm/lib/extensions.js:67:47)
      at dbFactory (src/database/db.js:39:135)
      at Object.done (src/ethereum/nation.test.js:369:41)
          at new Promise (<anonymous>)
          at <anonymous>

  ● nation › submitDraft › success

    Operation not permitted

      at Function.open (node_modules/realm/lib/extensions.js:67:47)
      at dbFactory (src/database/db.js:39:135)
      at Object.done (src/ethereum/nation.test.js:417:41)
          at new Promise (<anonymous>)
          at <anonymous>

  ● nation › saveAndSubmit › success

    Operation not permitted

      at Function.open (node_modules/realm/lib/extensions.js:67:47)
      at dbFactory (src/database/db.js:39:135)
      at Object.done (src/ethereum/nation.test.js:445:41)
          at new Promise (<anonymous>)
          at <anonymous>

  ● nation › deleteDraft › success

    Operation not permitted

      at Function.open (node_modules/realm/lib/extensions.js:67:47)
      at dbFactory (src/database/db.js:39:135)
      at Object.done (src/ethereum/nation.test.js:502:41)
          at new Promise (<anonymous>)
          at <anonymous>

 FAIL  src/profile/profile.test.js
  ● Test suite failed to run

    TypeError: Cannot redefine property: Worker
        at Function.defineProperty (<anonymous>)

      at Object.<anonymous> (node_modules/realm/lib/index.js:125:14)
      at Object.<anonymous> (src/database/db.js:31:25)
      at Object.<anonymous> (src/profile/profile.test.js:5:11)
          at Generator.next (<anonymous>)
          at new Promise (<anonymous>)
          at Generator.next (<anonymous>)
          at <anonymous>
      at process._tickCallback (internal/process/next_tick.js:188:7)

 FAIL  src/ethereum/wallet.test.js
  ● Test suite failed to run

    TypeError: Cannot redefine property: Worker
        at Function.defineProperty (<anonymous>)

      at Object.<anonymous> (node_modules/realm/lib/index.js:125:14)
      at Object.<anonymous> (src/database/db.js:31:25)
      at Object.<anonymous> (src/ethereum/wallet.js:50:25)
      at Object.<anonymous> (src/ethereum/wallet.test.js:5:15)
          at Generator.next (<anonymous>)
          at new Promise (<anonymous>)
          at Generator.next (<anonymous>)
          at <anonymous>
      at process._tickCallback (internal/process/next_tick.js:188:7)

 FAIL  src/queues/transaction.test.js
  ● Test suite failed to run

    TypeError: Cannot redefine property: Worker
        at Function.defineProperty (<anonymous>)

      at Object.<anonymous> (node_modules/realm/lib/index.js:125:14)
      at Object.<anonymous> (src/queues/transaction.js:123:133)
      at Object.<anonymous> (src/queues/transaction.test.js:1:122)
          at Generator.next (<anonymous>)
          at new Promise (<anonymous>)
          at Generator.next (<anonymous>)
          at <anonymous>
      at process._tickCallback (internal/process/next_tick.js:188:7)

 FAIL  src/database/db.test.js
  ● Test suite failed to run

    TypeError: Cannot redefine property: Worker
        at Function.defineProperty (<anonymous>)

      at Object.<anonymous> (node_modules/realm/lib/index.js:125:14)
      at Object.<anonymous> (src/database/db.js:31:25)
      at Object.<anonymous> (src/database/db.test.js:3:11)
          at Generator.next (<anonymous>)
          at new Promise (<anonymous>)
          at Generator.next (<anonymous>)
          at <anonymous>
      at process._tickCallback (internal/process/next_tick.js:188:7)

 FAIL  src/queues/messaging.test.js
  ● Test suite failed to run

    TypeError: Cannot redefine property: Worker
        at Function.defineProperty (<anonymous>)

      at Object.<anonymous> (node_modules/realm/lib/index.js:125:14)
      at Object.<anonymous> (src/database/db.js:31:25)
      at Object.<anonymous> (src/queues/messaging.test.js:1:113)
          at Generator.next (<anonymous>)
          at new Promise (<anonymous>)
          at Generator.next (<anonymous>)
          at <anonymous>
      at process._tickCallback (internal/process/next_tick.js:188:7)

 PASS  src/ethereum/utils.test.js
 PASS  src/ethereum/PangeaProvider.test.js
 PASS  src/ethereum/web3.test.js
 FAIL  src/database/queries.test.js
  ● Test suite failed to run

    TypeError: Cannot redefine property: Worker
        at Function.defineProperty (<anonymous>)

      at Object.<anonymous> (node_modules/realm/lib/index.js:125:14)
      at Object.<anonymous> (src/database/queries.js:4:15)
      at Object.<anonymous> (src/database/queries.test.js:3:16)
          at Generator.next (<anonymous>)
          at new Promise (<anonymous>)
          at Generator.next (<anonymous>)
          at <anonymous>

 PASS  src/test_implementations/nodeJsSecureStorage.test.js

Test Suites: 7 failed, 4 passed, 11 total
Tests:       9 failed, 50 passed, 59 total
Snapshots:   0 total
Time:        3.908s, estimated 6s
Ran all test suites.
```

**Comment by Mnuzz** [comment]
I noticed in the linked issue (realm) that somebody commented that deleting and re-cloning solved the problem. I tried all of the above changes on a brand new re-clone (with the updated node and realm versions). However, no changes in the tests.

**Comment by Mnuzz** [comment]
I thought that perhaps my docker setup had too few system resources allocated to it. So I increased the CPUs from 2 to 8, and increased the ram from 2GB to 8GB. It was much faster to run the tests, but still the same failures. I will try running docker using the Mac as a host, to eliminate the possibility that it has something to do with a Windows host. From my understanding, the host should not matter, but I won't rule out the possibility that there are some system resources that behave differently, to the point of it being a show stopper. Re-running these steps on the Mac, will report findings.

**Comment by Mnuzz** [comment]
Different behavior on the Mac with the docker image. Notably, the absence of the "operation not permitted" :

```
 PASS  src/ethereum/utils.test.js
 PASS  src/ethereum/nation.test.js
 PASS  src/ethereum/wallet.test.js
 PASS  src/profile/profile.test.js
 FAIL  src/queues/transaction.test.js
  ● Test suite failed to run

    TypeError: Cannot redefine property: Worker
        at Function.defineProperty (<anonymous>)
      
      at Object.<anonymous> (node_modules/realm/lib/index.js:125:14)
      at Object.<anonymous> (src/queues/transaction.js:123:133)
      at Object.<anonymous> (src/queues/transaction.test.js:1:122)
          at Generator.next (<anonymous>)
          at new Promise (<anonymous>)
      at handle (node_modules/worker-farm/lib/child/index.js:44:8)
      at process.<anonymous> (node_modules/worker-farm/lib/child/index.js:51:3)
      at emitTwo (events.js:126:13)
      at process.emit (events.js:214:7)
      at emit (internal/child_process.js:772:12)
      at _combinedTickCallback (internal/process/next_tick.js:141:11)
      at process._tickCallback (internal/process/next_tick.js:180:9)

 FAIL  src/queues/messaging.test.js
  ● Test suite failed to run

    TypeError: Cannot redefine property: Worker
        at Function.defineProperty (<anonymous>)
      
      at Object.<anonymous> (node_modules/realm/lib/index.js:125:14)
      at Object.<anonymous> (src/database/db.js:31:25)
      at Object.<anonymous> (src/queues/messaging.test.js:1:113)
          at Generator.next (<anonymous>)
          at new Promise (<anonymous>)
      at handle (node_modules/worker-farm/lib/child/index.js:44:8)
      at process.<anonymous> (node_modules/worker-farm/lib/child/index.js:51:3)
      at emitTwo (events.js:126:13)
      at process.emit (events.js:214:7)
      at emit (internal/child_process.js:772:12)
      at _combinedTickCallback (internal/process/next_tick.js:141:11)
      at process._tickCallback (internal/process/next_tick.js:180:9)

 PASS  src/test_implementations/nodeJsSecureStorage.test.js
 FAIL  src/database/db.test.js
  ● Test suite failed to run

    TypeError: Cannot redefine property: Worker
        at Function.defineProperty (<anonymous>)
      
      at Object.<anonymous> (node_modules/realm/lib/index.js:125:14)
      at Object.<anonymous> (src/database/db.js:31:25)
      at Object.<anonymous> (src/database/db.test.js:3:11)
          at Generator.next (<anonymous>)
          at new Promise (<anonymous>)
      at handle (node_modules/worker-farm/lib/child/index.js:44:8)
      at process.<anonymous> (node_modules/worker-farm/lib/child/index.js:51:3)
      at emitTwo (events.js:126:13)
      at process.emit (events.js:214:7)
      at emit (internal/child_process.js:772:12)
      at _combinedTickCallback (internal/process/next_tick.js:141:11)
      at process._tickCallback (internal/process/next_tick.js:180:9)

 PASS  src/ethereum/PangeaProvider.test.js
 FAIL  src/database/queries.test.js
  ● Test suite failed to run

    TypeError: Cannot redefine property: Worker
        at Function.defineProperty (<anonymous>)
      
      at Object.<anonymous> (node_modules/realm/lib/index.js:125:14)
      at Object.<anonymous> (src/database/queries.js:4:15)
      at Object.<anonymous> (src/database/queries.test.js:3:16)
          at Generator.next (<anonymous>)
          at new Promise (<anonymous>)
      at handle (node_modules/worker-farm/lib/child/index.js:44:8)
      at process.<anonymous> (node_modules/worker-farm/lib/child/index.js:51:3)
      at emitTwo (events.js:126:13)
      at process.emit (events.js:214:7)
      at emit (internal/child_process.js:772:12)
      at _combinedTickCallback (internal/process/next_tick.js:141:11)
      at process._tickCallback (internal/process/next_tick.js:180:9)

 PASS  src/ethereum/web3.test.js

Test Suites: 4 failed, 7 passed, 11 total
Tests:       78 passed, 78 total
Snapshots:   0 total
Time:        7.163s
```

**Comment by Mnuzz** [comment]
With the Mac, I did a full re-clone within the docker image, to the home directory of the image (not the mounted /app), so the file system itself is hosted within Docker. Still the same errors. However, doing the same on Windows got rid of the "operation not permitted" errors, with 6 Test Suites failing with "Cannot redefine property: Worker" (The Mac-hosted Docker image had 4 such Suites failing with that same error).

**Comment by Mnuzz** [comment]
Made a big breakthrough finding. This has to do with the Singleton object being re-constructed multiple times, whereas it is only meant to be constructed once per lifecycle of the process. Unit tests will often re-create the same objects within the same process.

Running the tests one at a time will not throw this error. So this confirms what I had believed. I should be able to come up with a solid fix under this principle...

**Comment by Mnuzz** [comment]
I checked the code of the realm/lib/index.js files between the working (mac) and non-working (docker) versions. Different code, despite package.lock.json showing the correct version. So I looked in the docker image's `~/.npm/` cache for realm. Guess what? It's showing version 2.2.7, and it's supposed to be using 2.1.1. 

I googled `npm cache wrong version` and found this: https://github.com/npm/npm/issues/12654

Version 3.x of npm, which is tied to the 6.x version of node, has this problem. 

The workaround (npm install twice) did not work for me. However, changing the docker image to use 8.9.4 of node allowed the correct version of realm to be installed, and therefore the issue was solved.

![image](https://user-images.githubusercontent.com/1613498/36057991-2f8212f2-0dd4-11e8-9896-2836efb2799e.png)
